File: src/core/execution/node_base.py
"""
Base node class for execution engine
"""
from abc import ABC, abstractmethod
from typing import Dict, Any, Optional, List, TYPE_CHECKING
from dataclasses import dataclass, field
from enum import Enum
import copy
from ..types import NodeID, NodeData

if TYPE_CHECKING:
    from ..types import NodeGraph


class ExecutionMode(Enum):
    """
    Node execution modes for autonomous workflows
    
    ONCE: Execute once per workflow run (default behavior)
    CONTINUOUS: Keep executing on each tick (scheduler nodes)
    TRIGGERED: Execute only when trigger input fires
    """
    ONCE = "once"
    CONTINUOUS = "continuous"
    TRIGGERED = "triggered"


@dataclass
class ExecutionContext:
    """Context passed to nodes during execution"""
    container: Any = None  # Optional ServiceContainer (for backward compatibility, nodes should be self-contained)
    variables: Dict[str, Any] = field(default_factory=dict)  # Runtime variables (user_id, user_query, etc.)
    node_outputs: Dict[NodeID, Dict[str, Any]] = field(default_factory=dict)  # Cached outputs from previous nodes


class BaseNode(ABC):
    """
    Base class for all execution nodes
    
    Each node:
    - Has a unique ID
    - Defines inputs and outputs
    - Executes when all dependencies are met
    - Can access ServiceContainer and execution context
    """
    
    # Default execution mode for the node class (override in subclasses)
    execution_mode: ExecutionMode = ExecutionMode.ONCE
    
    def __init__(self, node_id: NodeID, node_data: NodeData):
        """
        Initialize node
        
        Args:
            node_id: Unique node identifier
            node_data: Node data from workflow JSON
        """
        self.node_id = node_id
        self.node_data = node_data
        self.node_type = node_data.get('type', '')
        # Deep copy inputs to prevent mutations from affecting original workflow
        self.inputs = copy.deepcopy(node_data.get('inputs', {}))
        self.position = node_data.get('position', {'x': 0, 'y': 0})
        self.metadata = node_data.get('metadata', {})
        # State for triggered nodes
        self._triggered = False
        self._last_trigger_value = None
    
    @abstractmethod
    def execute(self, context: ExecutionContext) -> Dict[str, Any]:
        """
        Execute the node
        
        Args:
            context: Execution context with container, variables, and node outputs
            
        Returns:
            Dictionary of output values (keys match output names)
        """
        pass
    
    def get_input_value(self, input_name: str, context: ExecutionContext, default: Any = None) -> Any:
        """
        Get input value, resolving connections from other nodes
        
        Args:
            input_name: Name of the input
            context: Execution context
            default: Default value if not found
            
        Returns:
            Resolved input value
        """
        # Check if input is connected to another node's output
        # For now, we'll resolve this in the engine
        # This method can be overridden for special handling
        
        # First check if it's a direct value in inputs
        if input_name in self.inputs:
            value = self.inputs[input_name]
            
            # Check if it's a template variable (e.g., "{{user_query}}")
            if isinstance(value, str) and value.startswith('{{') and value.endswith('}}'):
                var_name = value[2:-2].strip()
                return context.variables.get(var_name, default)
            
            return value
        
        return default
    
    def get_connected_input(self, input_name: str, context: ExecutionContext) -> Optional[Any]:
        """
        Get input value from connected node output
        
        Args:
            input_name: Name of the input
            context: Execution context with node outputs
            
        Returns:
            Value from connected node, or None if not connected
        """
        # This will be resolved by the engine based on connections
        # For now, return None - engine handles connection resolution
        return None
    
    def initialize(self, workflow: 'NodeGraph', all_nodes: Dict[NodeID, 'BaseNode']) -> None:
        """
        Initialize node after all nodes are built
        Called by engine to allow nodes to set up relationships if needed.
        
        Args:
            workflow: Workflow definition with nodes and connections
            all_nodes: Dictionary of all node instances (node_id -> node)
        """
        # Default implementation does nothing
        # Override in subclasses for custom initialization
        pass
    
    def _setup(self, workflow: 'NodeGraph', all_nodes: Dict[NodeID, 'BaseNode']) -> None:
        """
        Setup node after all nodes are built (alias for initialize for backward compatibility)
        Called by engine to allow nodes to set up relationships if needed.
        
        Args:
            workflow: Workflow definition with nodes and connections
            all_nodes: Dictionary of all node instances (node_id -> node)
        """
        # Call initialize for backward compatibility
        self.initialize(workflow, all_nodes)
    
    def is_autonomous(self) -> bool:
        """
        Check if this node runs autonomously (CONTINUOUS mode)
        
        Returns:
            True if node has CONTINUOUS execution mode
        """
        return self.execution_mode == ExecutionMode.CONTINUOUS
    
    def is_triggered(self) -> bool:
        """
        Check if this node is trigger-based (TRIGGERED mode)
        
        Returns:
            True if node has TRIGGERED execution mode
        """
        return self.execution_mode == ExecutionMode.TRIGGERED
    
    def set_triggered(self, value: bool = True) -> None:
        """
        Set the triggered state for this node
        
        Args:
            value: Whether the node has been triggered
        """
        self._triggered = value
    
    def check_and_clear_trigger(self) -> bool:
        """
        Check if node was triggered and clear the trigger state
        
        Returns:
            True if node was triggered (and clears the state)
        """
        was_triggered = self._triggered
        self._triggered = False
        return was_triggered
    
    def on_tick(self, context: ExecutionContext) -> Optional[Dict[str, Any]]:
        """
        Called on each tick for CONTINUOUS nodes
        Override in subclasses to implement tick behavior.
        
        Args:
            context: Execution context
            
        Returns:
            Output dict if node should fire this tick, None otherwise
        """
        # Default implementation does nothing
        # Override in CONTINUOUS nodes (like SchedulerNode)
        return None

File: ts/src/core/execution/nodeBase.ts
/**
 * Base node class and execution context.
 * Mirrors Python src/core/execution/node_base.py
 */
import { NodeID, NodeData, StorageInterface, WorkflowData } from "../types";
import { getLogger } from "../../utils/logger";

const logger = getLogger("nodeBase");

/**
 * Node execution modes for autonomous workflows.
 *
 * ONCE:       Execute once per workflow run (default behaviour)
 * CONTINUOUS: Keep executing on each tick (scheduler / listener nodes)
 * TRIGGERED:  Execute only when a trigger input fires
 */
export enum ExecutionMode {
  ONCE = "once",
  CONTINUOUS = "continuous",
  TRIGGERED = "triggered",
}

/** Shared context passed to every node during execution */
export interface ExecutionContext {
  /** Context variables (user_query, user_id, etc.) */
  variables: Record<string, unknown>;
  /** Collected outputs from already-executed upstream nodes, keyed by nodeId */
  nodeOutputs: Record<NodeID, Record<string, unknown>>;
  /** Active storage instance (resolved from MemoryStorageNode) */
  storage?: StorageInterface;
}

/**
 * Abstract base class for all workflow nodes.
 *
 * Subclasses MUST override `execute()`.
 */
export abstract class BaseNode {
  readonly nodeId: string;
  readonly nodeType: string;
  inputs: Record<string, unknown>;
  metadata: Record<string, unknown>;
  position: { x: number; y: number };

  /** Default execution mode — override in subclasses (e.g. SchedulerNode, TelegramListenerNode). */
  static executionMode: ExecutionMode = ExecutionMode.ONCE;

  /** Populated by the engine before execution: maps inputName → {nodeId, outputName} */
  inputConnections: Record<string, { nodeId: string; outputName: string }[]> =
    {};

  /** Internal triggered state (for TRIGGERED mode nodes) */
  private _triggered = false;

  constructor(nodeId: string, nodeData: NodeData) {
    this.nodeId = nodeId;
    this.nodeType = nodeData.type;
    // Deep copy inputs to prevent mutations from affecting original workflow
    this.inputs = JSON.parse(
      JSON.stringify((nodeData.inputs as Record<string, unknown>) ?? {})
    );
    this.metadata = (nodeData.metadata as Record<string, unknown>) ?? {};
    this.position = nodeData.position ?? { x: 0, y: 0 };
  }

  // ── Setup / Initialization ──────────────────────────────────────────

  /**
   * Initialize node after all nodes are built.
   * Called by engine/runner to allow nodes to set up relationships and state.
   * Override in subclasses for custom initialization (e.g. TelegramListenerNode
   * fetches bot info, SchedulerNode seeds timing state).
   *
   * May return a Promise for async initialisation (e.g. network calls).
   *
   * Mirrors Python `initialize()`.
   */
  initialize(
    _workflow: WorkflowData,
    _allNodes: Map<NodeID, BaseNode>
  ): void | Promise<void> {
    // Default implementation does nothing
  }

  /**
   * Setup node after all nodes are built (alias for initialize for backward compat).
   * Called by engine to allow nodes to set up relationships.
   *
   * Mirrors Python `_setup()`.
   */
  _setup(
    workflow: WorkflowData,
    allNodes: Map<NodeID, BaseNode>
  ): void | Promise<void> {
    return this.initialize(workflow, allNodes);
  }

  // ── Autonomous-node helpers ───────────────────────────────────────

  /** True when this node runs continuously (CONTINUOUS mode). */
  isAutonomous(): boolean {
    return (this.constructor as typeof BaseNode).executionMode === ExecutionMode.CONTINUOUS;
  }

  /** True when this node only runs when triggered. */
  isTriggered(): boolean {
    return (this.constructor as typeof BaseNode).executionMode === ExecutionMode.TRIGGERED;
  }

  /** Set the triggered state for this node. */
  setTriggered(value = true): void {
    this._triggered = value;
  }

  /** Check if node was triggered and clear the trigger state. */
  checkAndClearTrigger(): boolean {
    const wasTriggered = this._triggered;
    this._triggered = false;
    return wasTriggered;
  }

  /**
   * Called on each runner tick for CONTINUOUS nodes.
   * Return an output dict to "fire" the node this tick, or `null` to skip.
   * Override in autonomous subclasses (SchedulerNode, TelegramListenerNode).
   */
  onTick(_context: ExecutionContext): Promise<Record<string, unknown> | null> | Record<string, unknown> | null {
    return null;
  }

  /** Execute the node logic. Must be implemented by subclasses. */
  abstract execute(
    context: ExecutionContext
  ): Promise<Record<string, unknown>> | Record<string, unknown>;

  /**
   * Clean up resources when the workflow is stopped (e.g. close HTTP server).
   * Override in subclasses that acquire resources in initialize().
   */
  dispose(): void | Promise<void> {
    // Default: no-op
  }

  // ── Input resolution ────────────────────────────────────────────────

  /**
   * Resolve an input value.
   *
   * Priority:
   * 1. Connected upstream output (via inputConnections)
   * 2. Direct value in this.inputs
   * 3. Metadata fallback (node.properties in the frontend)
   * 4. defaultValue
   */
  getInputValue(
    inputName: string,
    context: ExecutionContext,
    defaultValue: unknown = undefined
  ): unknown {
    // 1. Check connections
    const connections = this.inputConnections[inputName];
    if (connections && connections.length > 0) {
      // Use the first connection (most recent)
      const conn = connections[0];
      const upstreamOutputs = context.nodeOutputs[conn.nodeId];
      if (upstreamOutputs !== undefined) {
        const val = upstreamOutputs[conn.outputName];
        if (val !== undefined) {
          logger.debug(
            `[Node ${this.nodeId}] Input '${inputName}' resolved from node ${conn.nodeId}.${conn.outputName}`
          );
          return val;
        }
      }
    }

    // 2. Direct input value (check if it's a template variable)
    if (inputName in this.inputs) {
      const raw = this.inputs[inputName];
      return this.resolveTemplateVariable(raw, context);
    }

    // 3. Metadata fallback
    if (inputName in this.metadata) {
      const raw = this.metadata[inputName];
      return this.resolveTemplateVariable(raw, context);
    }

    return defaultValue;
  }

  // ── Template / Env variable resolution ──────────────────────────────

  /**
   * Resolve {{process.env.VAR}} templates against real environment variables.
   * Works without an ExecutionContext — safe to call for metadata values.
   */
  protected resolveEnvVar(value: unknown): unknown {
    if (typeof value !== "string") return value;

    // Single template: {{process.env.VAR}}
    const fullMatch = value.match(/^\{\{(.+?)\}\}$/);
    if (fullMatch) {
      const varName = fullMatch[1].trim();
      if (varName.startsWith("process.env.")) {
        const envKey = varName.slice("process.env.".length);
        return process.env[envKey] ?? value;
      }
    }

    // Inline replacement for multiple templates
    return value.replace(/\{\{(.+?)\}\}/g, (_match, varName: string) => {
      const trimmed = varName.trim();
      if (trimmed.startsWith("process.env.")) {
        const envKey = trimmed.slice("process.env.".length);
        return process.env[envKey] ?? _match;
      }
      return _match;
    });
  }

  /** Resolve {{varName}} templates against context.variables and process.env */
  protected resolveTemplateVariable(
    value: unknown,
    context: ExecutionContext
  ): unknown {
    if (typeof value !== "string") return value;

    // If the entire value is a single template, return the raw variable
    // (preserves non-string types like objects / numbers).
    const fullMatch = value.match(/^\{\{(.+?)\}\}$/);
    if (fullMatch) {
      const varName = fullMatch[1].trim();
      // Check process.env first
      if (varName.startsWith("process.env.")) {
        const envKey = varName.slice("process.env.".length);
        return process.env[envKey] ?? value;
      }
      // Only resolve if variable exists in context (don't overwrite with undefined)
      if (varName in context.variables) {
        return context.variables[varName];
      }
      return value; // leave unresolved
    }

    // Otherwise replace all {{var}} occurrences inline (always returns string).
    return value.replace(/\{\{(.+?)\}\}/g, (_match, varName: string) => {
      const trimmed = varName.trim();
      if (trimmed.startsWith("process.env.")) {
        const envKey = trimmed.slice("process.env.".length);
        return process.env[envKey] ?? _match;
      }
      const resolved = context.variables[trimmed];
      return resolved !== undefined ? String(resolved) : _match;
    });
  }
}
